File: gcpc2024/musicalmending/submissions/wrong_answer/ternary.java
import java.util.Scanner;

class ternary {
	int n;
	int notes[];
	
	long tune(int ref) {
		long res = 0;
		for (int i = 0; i < n; ++i) {
			res += Math.abs(ref - notes[i] + i);
		}
		
		// minimize function
		return -res;
	}
	
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);

		ternary t = new ternary();
		t.n = sc.nextInt();

		t.notes = new int[t.n];
		for(int i = 0; i < t.n; ++i) {
			t.notes[i] = sc.nextInt();
		}

		int l = -200010;
		int r = 200010;

		while (r - l > 1) {
			int mid = (r + l) / 2;
			
			if (t.tune(mid) < t.tune(mid + 1)) {
				l = mid;
			} else {
				r = mid;
			}
		}

		System.out.println(-t.tune(r));
}
}

File: gcpc2024/musicalmending/submissions/accepted/ternary.cpp
#include <iostream>
#include <vector>
#include <cassert>

using namespace std;

int main() {
	int n; cin >> n;

	vector<int> notes(n);
	for(int i = 0; i < n; ++i) {
		cin >> notes[i];
	}

	auto tune = [&](int ref) {
		long res = 0;
		for (int i = 0; i < n; ++i) {
			res += abs(ref - notes[i] + i);
		}

		// minimize function
		return -res;
	};

	int l = -300000;
	int r = 200000;

	while (r - l > 1) {
		int mid = (r + l) / 2;

		if (tune(mid) < tune(mid + 1)) {
			l = mid;
		} else {
			r = mid;
		}
	}

	cout << -tune(r) << endl;
}
